apps/omg/lib/omg/output.ex

```diff
@@ -19,56 +19,75 @@ defmodule OMG.Output do
 
   This module specificially dispatches generic calls to the various specific types
   """
+  alias OMG.Crypto
+
+  @type t :: %__MODULE__{
+          output_type: binary(),
+          owner: Crypto.address_t(),
+          currency: Crypto.address_t(),
+          amount: non_neg_integer()
+        }
+
+  defstruct [:output_type, :owner, :currency, :amount]
 
   alias OMG.RawData
+  alias OMG.Utxo
+
+  require Utxo
 
   @output_types_modules OMG.WireFormatTypes.output_type_modules()
   @output_types Map.keys(@output_types_modules)
 
-  def dispatching_reconstruct([raw_type, rlp_decoded_chunks]) when is_binary(raw_type) do
+  def reconstruct([raw_type, rlp_decoded_chunks]) when is_binary(raw_type) do
     case RawData.parse_uint256(raw_type) do
       {:ok, output_type} when output_type in @output_types ->
-        protocol_module = @output_types_modules[output_type]
-        protocol_module.reconstruct([output_type, rlp_decoded_chunks])
+        reconstruct([output_type, rlp_decoded_chunks])
 
       {:ok, _unrecognized_type} ->
         {:error, :unrecognized_output_type}
     end
   end
 
-  def dispatching_reconstruct(_), do: {:error, :malformed_outputs}
+  @doc """
+  Reconstructs the structure from a list of RLP items
+  """
+  def reconstruct([output_type, [owner_rlp, currency_rlp, amount_rlp]]) do
+    with {:ok, cur12} <- RawData.parse_address(currency_rlp),
+         {:ok, owner} <- RawData.parse_address(owner_rlp),
+         :ok <- non_zero_owner(owner),
+         {:ok, amount} <- RawData.parse_amount(amount_rlp),
+         do: %__MODULE__{output_type: output_type, owner: owner, currency: cur12, amount: amount}
+  end
 
-  def from_db_value(%{type: output_type} = db_value), do: @output_types_modules[output_type].from_db_value(db_value)
-  # default clause for backwards compatibility
-  def from_db_value(%{} = db_value), do: OMG.Output.FungibleMoreVPToken.from_db_value(db_value)
-end
+  def reconstruct(_), do: {:error, :malformed_outputs}
 
-defprotocol OMG.Output.Protocol do
-  @moduledoc """
-  Captures the varying behavior of outputs that build the plasma chain
+  defp non_zero_owner(<<0::160>>), do: {:error, :output_guard_cant_be_zero}
+  defp non_zero_owner(_), do: :ok
 
-  Includes the "output predicate", within the `can_spend?/3` function
-  """
+  def dispatching_reconstruct(_), do: {:error, :malformed_outputs}
 
-  @doc """
-  True if a particular witness can unlock a particular output to be spent, given being put in a particular transaction
+  def from_db_value(%{type: output_type} = db_value), do: @output_types_modules[output_type].from_db_value(db_value)
 
-  Intended to be called in stateful validation
-  """
-  def can_spend?(output_spent, witness, raw_tx)
+  def from_db_value(%{owner: owner, currency: currency, amount: amount, output_type: output_type})
+      when is_binary(owner) and is_binary(currency) and is_integer(amount) and is_integer(output_type) do
```

Comment:
I don't understand these guards? who are they for? what are they protecting us from?